node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -939,10 +931,93 @@ impl AdvertisementError {
 		match self {
 			InvalidAssignment => Some(COST_WRONG_PARA),
 			RelayParentUnknown | UndeclaredCollator | Invalid(_) => Some(COST_UNEXPECTED_MESSAGE),
-			UnknownPeer |
-			ProtocolMismatch |
-			SecondedLimitReached |
-			ProspectiveParachainsUnavailable => None,
+			UnknownPeer | ProtocolMismatch | SecondedLimitReached => None,
+		}
+	}
+}
+
+// Requests backing to sanity check the advertisement.
+async fn can_second<Sender>(
+	sender: &mut Sender,
+	candidate_para_id: ParaId,
+	candidate_relay_parent: Hash,
+	candidate_hash: CandidateHash,
+	parent_head_data_hash: Hash,
+) -> bool
+where
+	Sender: CollatorProtocolSenderTrait,
+{
+	let request = CanSecondRequest {
+		candidate_para_id,
+		candidate_relay_parent,
+		candidate_hash,
+		parent_head_data_hash,
+	};
+	let (tx, rx) = oneshot::channel();
+	sender.send_message(CandidateBackingMessage::CanSecond(request, tx)).await;
+
+	rx.await.unwrap_or_else(|err| {
+		gum::warn!(
+			target: LOG_TARGET,
+			?err,
+			?candidate_relay_parent,
+			?candidate_para_id,
+			?candidate_hash,
+			"CanSecond-request responder was dropped",
+		);
+
+		false
+	})
+}
+
+/// Checks whether any of the advertisements are unblocked and attempts to fetch them.
+async fn request_unblocked_collations<Sender, I>(sender: &mut Sender, state: &mut State, blocked: I)
+where
+	Sender: CollatorProtocolSenderTrait,
+	I: IntoIterator<Item = ((ParaId, Hash), Vec<BlockedAdvertisement>)>,
+{
+	for (key, mut value) in blocked {
+		let (para_id, para_head) = key;
```

Comment:
This function needs to take ownership of the mapping in order to iterate over it because of borrowing rules.

Will be resolved with later with a refactoring of collator protocol